### DISCUSSION TITLE
Compatibility with NodeJS

### DISCUSSION THREAD

**Discussion by KaruroChori** [shadow/shadow#2432] [Q&A] [answered]
Is there any documented usage of shadow with nodejs applications?
In my case for example, the application is able to run up to a certain extent, but attempts to fetch data via http will hang the execution, unlike when run "properly".

The only issue I could observe are many instances of syscalls 125 (sys_capget) and 435 (cannot find it in any list) which are not implemented.

Edit: and one instance of 229 (clock_getres)

**Comment by stevenengler** [comment]
> Is there any documented usage of shadow with nodejs applications?

No we don't have any examples using it, and I'm not aware of anyone who has tried it yet. I don't see any reason why Shadow wouldn't support it, but I also don't know much about the internals of the nodejs runtime. We would like to support it and provide examples if we can.

> In my case for example, the application is able to run up to a certain extent, but attempts to fetch data via http will hang the execution, unlike when run "properly".

Can you post an example that reproduces this? For example the application, how to build it, and the shadow config you're running?

What do you mean by "will hang the execution"? Does shadow itself hang (does the shadow process ever end)? Or does the simulated application just not progress any further? Is there an `exitcode` file for the server (for example `shadow.data/hosts/server/server.node.1000.exitcode`), and what's its value?

> The only issue I could observe are many instances of syscalls 125 (sys_capget) and 435 (cannot find it in any list) which are not implemented.

Most applications will probably be able to handle a failed `capget()`, and for nodejs my guess is that it's probably coming from [`bool HasOnly(int capability)`](https://github.com/nodejs/node/blob/b6971602564fc93c536ad469947536b487c968ea/src/node_credentials.cc#L61). Syscall 435 is `clone3` which is usually called from libc, and libc will fall back to `clone` on systems that don't support it. For example, if you enable `--strace-logging-mode standard`, you'll probably see:

```
00:00:03.000 [tid 1000] syscall(435, ...) = -38 (ENOSYS: Function not implemented)
00:00:03.000 [tid 1000] clone(...) = 1001
```

> Edit: and one instance of 229 (clock_getres)

This one could be important, but it's impossible to know without being able to look at the application.

**Reply by KaruroChori** [reply]
Thank you for your reply.

>Can you post an example that reproduces this? For example the application, how to build it, and the shadow config you're running?

The minimal example to reproduce my issues is here: https://github.com/KaruroChori/base-blockchain
Once cloned
```
yarn
yarn run build
```

And `yarn run run` to test it outside shadow or `yarn run simulation-suite` to run the associated shadow simulation (task.yaml).

I am also attaching the result of the execution within shadow on my system.
[simulations.zip](https://github.com/shadow/shadow/files/9691362/simulations.zip)

>What do you mean by "will hang the execution"?

`FetchError: request to http://localhost:3000/register failed, reason: socket hang up`
As part of the error in stderr. The error code returned is just 1.

**Reply by sporksmith** [reply]
Do you know what it's trying to connect to on port 3000? If it's something running natively on the host instead of under Shadow, it won't work

**Reply by KaruroChori** [reply]
The same instance executed in node.js. It opens a route on port 3000 first and later (internally I think this is executed on a separate thread) it performs a request to "self".
Basically client and server are both children of the same base process in that demo and there is no crosstalk between different nodes spawned in the simulation.

**Comment by stevenengler** [comment, marked as answer]
There seems to be some issues running simple node.js applications in Shadow. Two examples are being tracked at #2451 and #2452. Hopefully the issues causing those to not work correctly are the same issues preventing your application from working, and fixing those will hopefully allow your application to run in Shadow.

**Reply by KaruroChori** [reply]
It seems like the test application is now working! Thank you for fixing it.
I will check if I have any further issue with the main project.

**Reply by KaruroChori** [reply]
I tested a slightly more complex version: instead of routing the request to localhost, it is directed to "seed", one of the nodes spawned. The hostname is recognized as valid (otherwhise an exception would be thrown) but the execution hangs again like it did before.
I updated the example repository https://github.com/KaruroChori/base-blockchain for reference.

**Reply by stevenengler** [reply]
The two issues above have been fixed, but we've also found #2457, which can cause the simulation to become stuck when a high logging level is used. So you should avoid using `--log-level trace` (and maybe `--log-level debug`) for now.

> I tested a slightly more complex version: instead of routing the request to localhost, it is directed to "seed", one of the nodes spawned.

The http-server example uses two different hosts and seems to work okay, so I'm not sure yet why yours isn't working.

**Reply by KaruroChori** [reply]
I also double-checked your code based on http-server and it works as expected.
However a similar simple example in fastify is not:
```
import Fastify from 'fastify'
const fastify = Fastify({
    logger: true
})

fastify.get('/', async (request, reply) => {
    reply.type('application/json').code(200)
    return { hello: 'world' }
})

fastify.listen({ port: 3000 }, (err, address) => {
    if (err) throw err
    // Server is now listening on ${address}
})
```
This is failing while executed in shadow by providing no reply to curl.
Fastify is built on top of two dependencies, [avvio](https://github.com/fastify/avvio) and the same http module you have been using in your previous trials. I made few quick tests to see if avvio was responsible, but its examples are executed correctly in shadow, which means my tests are inconclusive.

I think that virtually no one would use the http module directly, as writing servers with its native interface is a bit inconvenient. Most people would use ExpressJS, Koa or Fastify, so supporting node roughly equates to ensure applications based on these frameworks can run.

If there is anything I could do to help, please let me know :).


Edit:
I also tested [Koa](https://koajs.com/). Its basic examples are also working as expected.

**Reply by stevenengler** [reply]
> This is failing while executed in shadow by providing no reply to curl.

It looks like you're only binding on localhost. It works for me with:

```js
const fastify = require('fastify')({ logger: true })

fastify.get('/', async (request, reply) => {
    reply.type('application/json').code(200)
    return { hello: 'world' }
})

fastify.listen({ host: '0.0.0.0', port: 80 }, (err, address) => {
    if (err) throw err
})
```

```yaml
general:
  stop_time: 10s
  model_unblocked_syscall_latency: true

network:
  graph:
    type: 1_gbit_switch

hosts:
  server:
    network_node_id: 0
    processes:
    - path: /usr/bin/node
      args: ../../../server.js
      start_time: 3s
  client:
    network_node_id: 0
    processes:
    - path: /usr/bin/curl
      args: server
      start_time: 5s
```

```
$ rm -rf shadow.data && shadow --strace-logging-mode standard --log-level info shadow.yaml > shadow.log
$ cat shadow.data/hosts/client/client.curl.1000.stdout
{"hello":"world"}
```
